consensus/src/dag/dag_fetcher.rs

```diff
@@ -103,17 +105,71 @@ impl DagFetcher {
                 .and_then(FetchResponse::try_from)
                 .and_then(|response| response.verify(&remote_request, &self.epoch_state.verifier))
             {
+                let ceritified_nodes = response.certified_nodes();
                 // TODO: support chunk response or fallback to state sync
-                let mut dag_writer = self.dag.write();
-                for rounds in response.certified_nodes() {
-                    for node in rounds {
+                {
+                    let mut dag_writer = self.dag.write();
+                    for node in ceritified_nodes {
                         if let Err(e) = dag_writer.add_node(node) {
                             error!("Failed to add node {}", e);
                         }
                     }
                 }
-                local_request.notify();
+
+                if self.dag.read().all_exists(local_request.node().parents()) {
+                    local_request.notify();
+                } else {
+                    // TODO: implement retry logic
+                }
             }
         }
     }
 }
+
+#[derive(Debug, ThisError)]
+pub enum FetchHandleError {
+    #[error("target node is not present")]
+    TargetNotPresent,
+    #[error("causal parents not present")]
+    NotPresent,
+}
+
+pub struct FetchHandler {
+    dag: Arc<RwLock<Dag>>,
+}
+
+impl FetchHandler {
+    pub fn new(dag: Arc<RwLock<Dag>>) -> Self {
+        Self { dag }
+    }
+}
+
+impl RpcHandler for FetchHandler {
+    type Request = RemoteFetchRequest;
+    type Response = FetchResponse;
+
+    fn process(&mut self, message: Self::Request) -> anyhow::Result<Self::Response> {
+        let dag_reader = self.dag.read();
+
+        let target_node = dag_reader
+            .get_node(message.target())
+            .ok_or(FetchHandleError::TargetNotPresent)?;
```

Comment:
I think this is expected behavior when we fetch the parents of a certified node. It is possible that only a byzantine node will have the target. We can either:
1. Persist the node after we sign (today we only persist the signature).
2. Send parents metadata as part of the fetch request.